lib/src/rules/use_build_context_synchronously.dart

```diff
@@ -0,0 +1,220 @@
+    }
+  }
+
+  static bool inTestDir(CompilationUnit unit) {
+    var path = unit.declaredElement?.source.fullName;
+    return path != null && _testDirectories.any(path.contains);
+  }
+}
+
+class _AwaitVisitor extends RecursiveAstVisitor {
+  bool hasAwait = false;
+
+  @override
+  void visitAwaitExpression(AwaitExpression node) {
+    hasAwait = true;
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  static const _nameBuildContext = 'BuildContext';
+  final Uri _uriFramework;
+
+  final LintRule rule;
+
+  _Visitor(this.rule)
+      : _uriFramework = Uri.parse('package:flutter/src/widgets/framework.dart');
+
+  bool accessesContext(ArgumentList argumentList) {
+    for (var argument in argumentList.arguments) {
+      var argType = argument.staticType;
+      var isGetter = argument is Identifier &&
+          argument.staticElement is PropertyAccessorElement;
+      if (isBuildContext(argType, skipNullable: isGetter)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void check(AstNode node) {
+    // Walk back and look for an async gap that is not guarded by a mounted
+    // property check.
+    AstNode? child = node;
+    while (child != null && child is! FunctionBody) {
+      var parent = child.parent;
+      if (parent is Block) {
+        var statements = parent.statements;
+        var index = statements.indexOf(child as Statement);
+        for (var i = index - 1; i >= 0; i--) {
+          var s = statements.elementAt(i);
+          if (isMountedCheck(s)) {
+            return;
+          } else if (isAsync(s)) {
+            rule.reportLint(node);
+            return;
+          }
+        }
+      }
+
+      child = child.parent;
+    }
+  }
+
+  bool isAsync(Statement statement) {
+    var visitor = _AwaitVisitor();
+    statement.accept(visitor);
+    return visitor.hasAwait;
+  }
+
+  /// todo (pq): replace in favor of flutter_utils.isBuildContext
+  bool isBuildContext(DartType? type, {bool skipNullable = false}) {
+    if (type is! InterfaceType) {
+      return false;
+    }
+    if (skipNullable && type.nullabilitySuffix == NullabilitySuffix.question) {
+      return false;
+    }
+    var element = type.element;
+    return element.name == _nameBuildContext &&
+        element.source.uri == _uriFramework;
+  }
+
+  bool isMountedCheck(Statement statement) {
+    // This is intentionally naive.  Using a simple 'mounted' property check
+    // as a signal plays nicely w/ unanticipated framework classes that provide
+    // their own mounted checks.  The cost of this generality is the possibility
+    // of false negatives.
+    if (statement is IfStatement) {
+      var condition = statement.condition;
+      if (condition is PrefixExpression) {
+        if (condition.operator.type == TokenType.BANG) {
+          var operand = condition.operand;
+          // stateContext.mounted => mounted
+          while (operand is PrefixedIdentifier) {
+            operand = operand.identifier;
```

Comment:
`operand.identifier` will always return a `SimpleIdentifier`, so there's no need for a loop, just use an `if`.